torch/csrc/distributed/c10d/CUDASymmetricMemory.cu

```diff
@@ -398,50 +400,53 @@ void check_channel(int channel, int world_size) {
       ")");
 }
 
-__device__ __forceinline__ void release_signal(uint32_t* addr) {
-#if defined(USE_ROCM) || (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ < 800))
-  CUDA_KERNEL_ASSERT(false);
-#else
-  volatile uint32_t* signal = addr;
-  uint32_t val;
-  do {
-    val = *signal;
-  } while (val != 0 || atomicCAS_system(addr, 0, 1) != 0);
-#endif
-}
-
-__device__ __forceinline__ void acquire_signal(uint32_t* addr) {
-#if defined(USE_ROCM) || (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ < 800))
-  CUDA_KERNEL_ASSERT(false);
-#else
-  volatile uint32_t* signal = addr;
-  uint32_t val;
-  do {
-    val = *signal;
-  } while (val != 1 || atomicCAS_system(addr, 1, 0) != 1);
-#endif
-}
-
 static __global__ void barrier_kernel(
     uint32_t** signal_pads,
     int channel,
     int rank,
-    int world_size) {
+    int world_size,
+    size_t timeout_ms) {
   if (threadIdx.x < world_size) {
     auto target_rank = threadIdx.x;
-    release_signal(signal_pads[target_rank] + world_size * channel + rank);
-    acquire_signal(signal_pads[rank] + world_size * channel + target_rank);
+    if (target_rank == rank) {
+      return;
+    }
+    auto put_success = try_put_signal<MemOpSem::Release>(
+        signal_pads[target_rank] + world_size * channel + rank, timeout_ms);
+    if (!put_success) {
+      printf(
```

Comment:
I wonder, doesn't PyTorch have the "device-side assert" system, which is similar to CUDA's native device asserts but they are recoverable? Wouldn't it be a great fit for this?

Reply:
My impression is that it's not always enabled. I'll check if it's enabled in release builds and switch to it if it is.